tooling/nargo_fmt/src/formatter/statement.rs

```diff
@@ -0,0 +1,680 @@
+                    formatter.write_semicolon();
+                }));
+            }
+            StatementKind::Comptime(statement) => {
+                chunks.group(self.format_comptime_statement(*statement));
+            }
+            StatementKind::Semi(expression) => {
+                chunks.group(self.format_semi_statement(expression));
+            }
+            StatementKind::Interned(..) | StatementKind::Error => {
+                unreachable!("Should not be present in the AST")
+            }
+        }
+    }
+
+    fn format_let_statement(&mut self, let_statement: LetStatement) -> ChunkGroup {
+        self.format_let_or_global(
+            Keyword::Let,
+            let_statement.pattern,
+            let_statement.r#type,
+            Some(let_statement.expression),
+            let_statement.attributes,
+        )
+    }
+
+    pub(super) fn format_let_or_global(
+        &mut self,
+        keyword: Keyword,
+        pattern: Pattern,
+        typ: UnresolvedType,
+        value: Option<Expression>,
+        attributes: Vec<SecondaryAttribute>,
+    ) -> ChunkGroup {
+        let mut group = ChunkGroup::new();
+
+        group.text(self.chunk(|formatter| {
+            if !attributes.is_empty() {
+                formatter.format_attributes();
+            }
+            formatter.write_keyword(keyword);
+            formatter.write_space();
+            formatter.format_pattern(pattern);
+            if typ.typ != UnresolvedTypeData::Unspecified {
+                formatter.write_token(Token::Colon);
+                formatter.write_space();
+                formatter.format_type(typ);
+            }
+        }));
+
+        if let Some(value) = value {
+            group.text(self.chunk(|formatter| {
+                formatter.write_space();
+                formatter.write_token(Token::Assign);
+                formatter.write_space();
+            }));
+
+            let mut value_group = ChunkGroup::new();
+            value_group.kind = GroupKind::AssignValue;
+            self.format_expression(value, &mut value_group);
+            value_group.semicolon(self);
+            group.group(value_group);
+        } else {
+            group.semicolon(self);
+        }
+
+        group
+    }
+
+    fn format_constrain_statement(
+        &mut self,
+        constrain_statement: ConstrainStatement,
+    ) -> ChunkGroup {
+        let mut group = ChunkGroup::new();
+
+        let keyword = match constrain_statement.kind {
+            ConstrainKind::Assert => Keyword::Assert,
+            ConstrainKind::AssertEq => Keyword::AssertEq,
+            ConstrainKind::Constrain => {
+                unreachable!("constrain always produces an error, and the formatter doesn't run when there are errors")
+            }
+        };
+
+        group.text(self.chunk(|formatter| {
+            formatter.write_keyword(keyword);
+            formatter.write_left_paren();
+        }));
```

Comment:
How do you not forget whether to call things in `self.chunk` or directly like `self.write_left_paren()`?

Reply:
Do you see a way to enforce when the formatter can be written to?

Reply:
> How do you not forget whether to call things in self.chunk or directly like self.write_left_paren()?

It seems that Rust gives an error if you try to use `self` there (it says "cannot borrow `*self` as mutable because previous closure requires unique access second borrow occurs here")

Reply:
Sorry, my question wasn't clear. I didn't mean how you remember not to call `self.write_left_paren()` inside the lambda passed to `self.chunk` (which is where the "cannot borrow as mutable" error comes up), but rather how you remember to do this:

```
        group.text(self.chunk(|formatter| {
            formatter.write_keyword(keyword);
            formatter.write_left_paren();
        }));
```
instead of:
```
          self.write_keyword(keyword);
          self.write_left_paren();
```

The compiler wouldn't complain if `group.whatever(...)` is mixed with `self.write_something()`. This is where I find the fact that some `format_*` methods return a `ChunkGroup` while others take a mutable reference to them. 

I was wondering for example if instead of `Formatter::write_*` we could have a `Writer` that is only passed to the lambda by `Formatter::chunk`, so in a `format_*` method we don't have to wonder if we have to call `write_*`, or call `chunk`, because to write we _must_ call `chunk`. But I don't know if this is realistic, just wanted to point out that sometimes I only see that `.chunk` needs to be used because `ChunkGroup` is returned, so calling `self.write` couldn't possibly make sense, but other times it's not returned and it might be harder to see what the right thing to do is.

Reply:
Ah, I understand now.

Right now the formatter has two mixed modes: format without accounting for the maximum width (this is done for the first line of a module declaration, trait and impl) or accounting for it (statements and expressions). If the maximum width is not important (or unlikely to be reached) we use `self.write_keyword(...)` as that writes directly to the buffer. Otherwise we use groups and chunks because formatting those will check for the maximum width.

Maybe everything would be more consistent if it was always composed of groups and chunks, but it would make the code slightly more complex. We could try, though! (not sure if this in PR or as a follow-up)

Reply:
Aha, thanks for that explanation. It would be nice to be able to express this distinction in code, that some formatters require a writer that doesn't deal with chunks, while others need to construct chunks before the real writing begins, but I see that it's a bit intertwined, like `write_token` also calls `skip_comments_and_whitespace`, and I don't have a good idea now how to do it differently. I thought about using a phantom type, but don't think it would work with `&mut self`.

Reply:
It's a good point. Sometimes I would write a test to see it fail, then try to implement something and it was still failing... because I was writing directly to the buffer instead of creating a chunk and appending it to an existing group. Luckily the test caught that. I think we could refactor the code to make it more "type safe" in that regards, but with the many tests we have maybe it's good enough: if you use it the wrong way a test will fail.

Reply:
Here's an idea that is might be relatively easy to implement without radical refactoring: https://gist.github.com/aakoshh/a813dd5291fc48645d0f07b7a894b8b0

It's based on wrapping the mutable reference and implementing formatters on the wrappers.